modules/cats/src/smithy4s/interopcats/SchemaVisitorHash.scala

```diff
@@ -0,0 +1,177 @@
+package smithy4s.interopcats
+
+import cats.Hash
+import cats.implicits.{catsKernelStdHashForList, catsKernelStdHashForOption, toContravariantOps}
+
+import smithy4s.{Bijection, Hints, Lazy, Refinement, ShapeId}
+import smithy4s.capability.EncoderK
+import smithy4s.interopcats.instances.HashInstances._
+import smithy4s.schema.{Alt, CollectionTag, EnumValue, Field, Primitive, Schema, SchemaAlt, SchemaField, SchemaVisitor}
+
+
+
+object SchemaVisitorHash extends SchemaVisitor[Hash] { self =>
+
+  override def primitive[P](
+      shapeId: ShapeId,
+      hints: Hints,
+      tag: Primitive[P]
+  ): Hash[P] = primHashPf(tag)
+
+  override def collection[C[_], A](
+      shapeId: ShapeId,
+      hints: Hints,
+      tag: CollectionTag[C],
+      member: Schema[A]
+  ): Hash[C[A]] = {
+    implicit val memberHash: Hash[A] = self(member)
+    tag match {
+      case CollectionTag.ListTag       => Hash[List[A]]
+      case CollectionTag.SetTag        => Hash[Set[A]]
+      case CollectionTag.VectorTag     => Hash[Vector[A]]
+      case CollectionTag.IndexedSeqTag => Hash[Seq[A]].contramap(_.toIndexedSeq)
+    }
+  }
+
+  override def map[K, V](
+      shapeId: ShapeId,
+      hints: Hints,
+      key: Schema[K],
+      value: Schema[V]
+  ): Hash[Map[K, V]] = {
+    implicit val keyHash: Hash[K] = self(key)
+    implicit val valueHash: Hash[V] = self(value)
+    Hash[Map[K, V]]
+  }
+
+  override def enumeration[E](
+      shapeId: ShapeId,
+      hints: Hints,
+      values: List[EnumValue[E]],
+      total: E => EnumValue[E]
+  ): Hash[E] = {
+    implicit val enumValueHash: Hash[EnumValue[E]] = Hash[String].contramap(_.stringValue)
+    Hash[EnumValue[E]].contramap(total)
+  }
+
+  override def struct[S](
+      shapeId: ShapeId,
+      hints: Hints,
+      fields: Vector[SchemaField[S, _]],
+      make: IndexedSeq[Any] => S
+  ): Hash[S] = {
+    new Hash[S] {
+      override def hash(x: S): Int = {
+        def forField[A2](field: Field[Schema, S, A2]): Hash[S] = {
+          val hashField: Hash[A2] =
+            field.foldK(new Field.FolderK[Schema, S, Hash]() {
+              override def onRequired[A](
+                  label: String,
+                  instance: Schema[A],
+                  get: S => A
+              ): Hash[A] = self(instance)
+
+              override def onOptional[A](
+                  label: String,
+                  instance: Schema[A],
+                  get: S => Option[A]
+              ): Hash[Option[A]] = {
+                implicit val hashA: Hash[A] = self(instance)
+                Hash[Option[A]]
+              }
+            })
+          hashField.contramap(field.get)
+        }
+        fields.map(field => forField(field)).map(hash => hash.hash(x)).sum
```

Comment:
You're computing a visitor on the hot path. The `forField` should be called outside of the body of the `override def hash(x: S): Int` method.

Reply:
Moreover, the logic is off : you should call `combineHash` there.